### PR TITLE
Test-driven development (TDD)

### PR DESCRIPTION
**Pull request by CesarHerr**
### Hi there! @CesarHerr, @alexmedinasf and @Ruthmy here. 👋

This is the PR for the [TDD project](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/testing/tdd_project.md) activity. The goal of this milestone is to practice the TDD process by creating the tests for the methods that are needed to make the code work.

### Here is a summary of what has been done:

- [x] Separately, we created the tests for the methods that are needed to make the code work.
- [x] We created the methods that are needed to make the tests pass.
- [x] After the tests, we created a Solver class with three different methods: factorial, reverse and fizzbuzz.

---

### Thank you for taking the time to review this PR ⭐️!

---

### CODE REVIEW

File: spec/factorial_spec.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+    context 'returns 24 when given 4' do
+      let(:number) { 4 }
+      it { is_expected.to eq(24) }
+    end
+
+    context 'returns 120 when given 5' do
+      let(:number) { 5 }
+      it { is_expected.to eq(120) }
+    end
```
Comment (usorfaitheloho) [inline spec/factorial_spec.rb:40]: You did well with your project, kindly note that your factorial should have test that handles negative value which raises an exception as an expectation

image

File: spec/fizzbuzz_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        expect(solver.fizzbuzz(divisible5)).to eq('buzz')
+      end
+
+      it 'returns "fizzbuzz" when divisible by both 3 and 5' do
+        # Act
+        divisible3_and5 = 15
+
+        # Assert
+        expect(solver.fizzbuzz(divisible3_and5)).to eq('fizzbuzz')
+      end
```
Comment (usorfaitheloho) [inline spec/fizzbuzz_spec.rb:31]: You didn't implement a test case for the last condition in the fizzbuzz test cases 👍 

image

**Review by usorfaitheloho — CHANGES_REQUESTED** [review]
> Hi @Ruthmy . @CesarHerr  and @alexmedinasf  ,
> 
> Good job so far! There are some issues that you still need to work on to go to the next project but you are almost there!

  [ ](https://camo.githubusercontent.com/679757134a8fee76cfe1941f9f3ecb0e0cf56dd02da4498268ce31f0dacdfca3/68747470733a2f2f6d65646961302e67697068792e636f6d2f6d656469612f556461535559436c723932636a57706556782f67697068792e676966)
> 
> ## Highlights
> * Created Solver class ✔️
> *Created a method called factorial ✔️
> * Used GitHub flow ✔️
> 
> ## Required Changes ♻️
> _Check the comments under the review._
> 
> ## Optional suggestions
> _Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
> 
> Cheers and Happy coding!👏👏👏
> 
> Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.
> 
> **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**
> 
> _As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/factorial_spec.rb (hunk 1 above)
Comment (Ruthmy) [reply spec/factorial_spec.rb:40]: Hi @usorfaitheloho ❗ 
Thank you very much for your feedback, you're totally right, I forgot to add that case. I will add it right away.
Thanks ✨

File: spec/fizzbuzz_spec.rb (hunk 2 above)
Comment (CesarHerr) [reply spec/fizzbuzz_spec.rb:31]: @usorfaitheloho Thanks for your feedback. We'll check it!

**Review by MahmoudBakr23 — APPROVED** [review]
Hi TEAM 😃, 

Your project is complete!  There is nothing else to say other than... It's time to merge it :shipit:
Congratulations! 🎉

**Highlights**
- Well done on making the required changes from the previous reviewer
- Excellent communication

Cheers, and Happy coding!👏👏👏

Feel free to mention @MahmoudBakr23 to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**